### PR TITLE
Make version parsing available

### PR DESCRIPTION
**Pull request by BMagnu**
Extracts the version parsing code from the Lua API.
Prerequisite for #5583

### CODE REVIEW

**Review by Goober5000 — COMMENTED** [review]
This is a good idea in principle, but there are now two version parsing functions -- this and `parse_version_inline`.  Can they be consolidated?

**Review by Goober5000 — COMMENTED** [review]
Huh, clever.

Is it possible for a version to be successfully parsed, yet have one of its numbers be -1?  It looks like there might need to be a check at the bottom of `version::version` to guard against this.

**Comment by BMagnu** [comment]
As is now, negative versions are allowed, but will prevent a revision to be parsed if one exists.
Previously, negative versions were also allowed, but would also allow a revision to be parsed.
I'll change it to allow for revision with negative versions

File: code/globalincs/version.cpp [hunk 1]
```diff
@@ -75,31 +75,45 @@ version::version(int major_in, int minor_in, int build_in, int revision_in) :
...
-				else if (build < 0) {
-					build = std::stoi(str.c_str());
-				}
-				str.erase(0, pos + 1);
-			}
-			else if (major < 0) {
-				break; //Break out of the loop if the first string is not a digit.
-			}
+			ver = str.substr(0, pos);
+			str.erase(0, pos + 1);
```
Comment (Goober5000) [inline code/globalincs/version.cpp:84]: With the new copy statement a few lines above, I don't think this statement here is correct

File: code/globalincs/version.cpp [hunk 2]
```diff
@@ -75,31 +75,45 @@ version::version(int major_in, int minor_in, int build_in, int revision_in) :
...
+			case 1:
+				minor = std::stoi(ver.c_str());
+				++has_fields;
+				break;
+			case 2:
+				build = std::stoi(ver.c_str());
+				++has_fields;
+				break;
+			case 3:
+				revision = std::stoi(str.c_str());
```
Comment (Goober5000) [inline code/globalincs/version.cpp:106]: Does this `str` also need to be `ver`?

Comment (BMagnu) [reply code/globalincs/version.cpp]: done

File: code/globalincs/version.cpp (hunk 1 above)
Comment (Goober5000) [reply code/globalincs/version.cpp:84]: Retracted - I misinterpreted what was going on

**Review by Goober5000 — APPROVED** [review]

**Pull request merged**